File: src-tauri/src/agent/contracts.rs
use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};

pub(crate) const MAX_SESSION_ID_LENGTH: usize = 512;

#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum AgentKind {
    Codex,
    Claude,
}

impl AgentKind {
    pub const ALL: [Self; 2] = [Self::Codex, Self::Claude];

    pub fn command_name(self) -> &'static str {
        match self {
            Self::Codex => "codex",
            Self::Claude => "claude",
        }
    }

    pub fn display_name(self) -> &'static str {
        match self {
            Self::Codex => "Codex",
            Self::Claude => "Claude Code",
        }
    }
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentCapabilities {
    pub launch: bool,
    pub resume: bool,
    pub history: bool,
    pub prompt: bool,
    pub structured_state: bool,
}

impl AgentCapabilities {
    pub fn cli_foundation() -> Self {
        Self {
            launch: true,
            resume: true,
            history: true,
            prompt: true,
            // Hooks are not wired yet; screen text remains an approximation.
            structured_state: false,
        }
    }
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentDescriptor {
    pub id: String,
    pub kind: AgentKind,
    pub display_name: String,
    pub command: String,
    pub capabilities: AgentCapabilities,
}

impl AgentDescriptor {
    pub fn for_kind(kind: AgentKind) -> Self {
        Self {
            id: format!("agent:{}", kind.command_name()),
            kind,
            display_name: kind.display_name().to_string(),
            command: kind.command_name().to_string(),
            capabilities: AgentCapabilities::cli_foundation(),
        }
    }
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentSessionRef {
    pub agent: AgentKind,
    pub id: String,
}

impl AgentSessionRef {
    pub fn validate(&self) -> Result<(), String> {
        validate_agent_session_id(&self.id)
    }
}

pub(crate) fn validate_agent_session_id(value: &str) -> Result<(), String> {
    if value.is_empty()
        || value == "."
        || value.len() > MAX_SESSION_ID_LENGTH
        || value.chars().any(char::is_control)
        || value.contains('/')
        || value.contains('\\')
        || value.contains("..")
    {
        return Err("agent session id is invalid".to_string());
    }
    Ok(())
}

#[allow(dead_code)]
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum AgentLifecycleState {
    Starting,
    Processing,
    AwaitingInput,
    Completed,
    Failed,
    Interrupted,
    Unknown,
}

#[allow(dead_code)]
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum AgentStateSource {
    Hook,
    ScreenHeuristic,
    Process,
}

#[allow(dead_code)]
#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentStateSnapshot {
    pub lifecycle: AgentLifecycleState,
    pub source: AgentStateSource,
    pub confidence: Option<f32>,
    pub reason: Option<String>,
}

#[allow(dead_code)]
#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentLifecycleEvent {
    pub schema_version: u16,
    pub agent: AgentKind,
    pub session: AgentSessionRef,
    pub state: AgentLifecycleState,
    pub source: AgentStateSource,
    pub occurred_at: i64,
    pub reason: Option<String>,
    pub metadata: Option<Map<String, Value>>,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentResumePlan {
    pub schema_version: u16,
    pub session: AgentSessionRef,
    pub supported: bool,
    pub arguments: Vec<String>,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentAvailability {
    pub descriptor: AgentDescriptor,
    pub kind: AgentKind,
    pub available: bool,
    pub executable: Option<String>,
    pub version: Option<String>,
    pub reason: Option<String>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn descriptor_serializes_the_stable_frontend_shape() {
        let value = serde_json::to_value(AgentDescriptor::for_kind(AgentKind::Claude)).unwrap();
        assert_eq!(value["id"], "agent:claude");
        assert_eq!(value["displayName"], "Claude Code");
        assert_eq!(value["capabilities"]["structuredState"], false);
    }

    #[test]
    fn session_refs_reject_path_like_ids_but_keep_opaque_ids() {
        let valid = AgentSessionRef {
            agent: AgentKind::Codex,
            id: "019ff0d5-dbaf-7893-96db-4fbbbfee03a7".to_string(),
        };
        assert!(valid.validate().is_ok());
        for id in ["", "../session", "a\\b", ".", "line\nbreak", "nul\0byte"] {
            let invalid = AgentSessionRef {
                agent: AgentKind::Codex,
                id: id.to_string(),
            };
            assert!(invalid.validate().is_err(), "{id} should be rejected");
        }
        let overlong_utf8 = AgentSessionRef {
            agent: AgentKind::Codex,
            id: "会".repeat(171),
        };
        assert!(overlong_utf8.validate().is_err());
    }
}

File: src/agent/contracts.ts
export type AgentKind = "codex" | "claude";

export type AgentLifecycleState =
  | "starting"
  | "processing"
  | "awaiting_input"
  | "completed"
  | "failed"
  | "interrupted"
  | "unknown";

export type AgentStateSource = "hook" | "screen_heuristic" | "process";

export interface AgentCapabilities {
  launch: boolean;
  resume: boolean;
  history: boolean;
  prompt: boolean;
  structuredState: boolean;
}

export interface AgentDescriptor {
  id: string;
  kind: AgentKind;
  displayName: string;
  command: string;
  capabilities: AgentCapabilities;
}

export interface AgentAvailability {
  descriptor: AgentDescriptor;
  kind: AgentKind;
  available: boolean;
  executable: string | null;
  version: string | null;
  reason: string | null;
}

export interface AgentSessionRef {
  agent: AgentKind;
  id: string;
}

export interface AgentStateSnapshot {
  lifecycle: AgentLifecycleState;
  source: AgentStateSource;
  confidence: number | null;
  reason: string | null;
}

export interface AgentLifecycleEvent {
  schemaVersion: number;
  agent: AgentKind;
  session: AgentSessionRef;
  state: AgentLifecycleState;
  source: AgentStateSource;
  occurredAt: number;
  reason: string | null;
  metadata: Record<string, unknown> | null;
}

export interface AgentResumePlan {
  schemaVersion: number;
  session: AgentSessionRef;
  supported: boolean;
  arguments: string[];
}

const MAX_SESSION_ID_LENGTH = 512;

export function isAgentKind(value: string): value is AgentKind {
  return value === "codex" || value === "claude";
}

export function isAgentSessionRef(value: unknown): value is AgentSessionRef {
  if (!isRecord(value) || !isAgentKindValue(value.agent) || typeof value.id !== "string") {
    return false;
  }
  return isAgentSessionId(value.id);
}

export function isAgentSessionId(value: string) {
  return value.length > 0
    && value !== "."
    && new TextEncoder().encode(value).byteLength <= MAX_SESSION_ID_LENGTH
    && !/[\u0000-\u001f\u007f-\u009f]/u.test(value)
    && !value.includes("/")
    && !value.includes("\\")
    && !value.includes("..");
}

export function agentProfileId(kind: AgentKind) {
  return `agent:${kind}` as const;
}

export function agentDescriptor(kind: AgentKind): AgentDescriptor {
  return {
    id: agentProfileId(kind),
    kind,
    displayName: kind === "codex" ? "Codex" : "Claude Code",
    command: kind,
    capabilities: {
      launch: true,
      resume: true,
      history: true,
      prompt: true,
      structuredState: false,
    },
  };
}

function isAgentKindValue(value: unknown): value is AgentKind {
  return typeof value === "string" && isAgentKind(value);
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}
